ftw/sliderblock/tests/test_dropzone_upload.py

```diff
@@ -0,0 +1,83 @@
+    #     self.assertEqual({u'content': u'Created',
+    #                       u'url': u'http://nohost/plone/page/downloads/world.txt',
+    #                       u'proceed': True}, browser.json)
+    #     self.assertEquals(['world.txt'], listing.objectIds())
+    #
+    #     doc, = listing.objectValues()
+    #     self.assertEquals('File', doc.portal_type)
+    #     self.assertEquals('world.txt', doc.Title())
+    #     if IS_PLONE_5:
+    #         self.assertEquals('Hello World', doc.file.data.strip())
+    #     else:
+    #         self.assertEquals('Hello World', doc.getFile().data.strip())
+
+    @browsing
+    def test_create_images_in_sliderblock(self, browser):
+        sliderblock = create(Builder('sliderblock').titled(u'Slider Block')
+                         .within(create(Builder('sl content page').titled(u'Page'))))
+        self.assertEquals([], sliderblock.objectIds())
+
+        browser.login()
+        self.make_dropzone_upload(sliderblock, self.asset('image.jpg').open('rb'))
+        self.assertEqual(201, browser.status_code)
+        self.assertEqual({u'content': u'Created',
+                          u'url': u'http://nohost/plone/page/sliderblock/cropped.jpg',
+                          u'proceed': True}, browser.json)
+        self.assertEquals(['cropped.jpg'], sliderblock.objectIds())
+
+        img, = sliderblock.objectValues()
+        self.assertEquals('Image', img.portal_type)
+        self.assertEquals('cropped.jpg', img.Title())
+
+    @browsing
+    def test_files_are_not_allowed_in_galleries(self, browser):
+        gallery = create(Builder('sl galleryblock').titled(u'Gallery')
+                         .within(create(Builder('sl content page').titled(u'Page'))))
+        self.assertEquals([], gallery.objectIds())
+
+        with browser.login().expect_http_error(400):
+            self.make_dropzone_upload(gallery, self.asset('world.txt').open('r'))
+
+        self.assertEqual({u'error': u'Only images can be added to the gallery.',
+                          u'proceed': False}, browser.json)
+        self.assertEquals([], gallery.objectIds())
+
+    @staticmethod
+    def make_dropzone_upload(context, file_):
+        """The dropzone JS makes a multipart upload with the the file as field named "file".
+        The testbrowser does not support making multipart requests directly (we would have to
+        create the MIME body ourself), but the form filling does that well.
+        Therefore we use an HTML form for uploading.
+        """
+        html = (
+            '<form action="{}/@@dropzone-upload" method="post" enctype="multipart/form-data">'
+            '  <input type="file" name="file" />'
+            '  <input type="hidden" name="_authenticator" value="{}" />'
+            '  <input type="submit" />'
+            '</form>').format(context.absolute_url(), createToken())
+        defaultbrowser.open(context).parse(html)
+        defaultbrowser.fill({'file': file_}).submit()
```

Comment:
I'm stuck here getting a permission denied error.

Reply:
If I understand the code correctly, the browser is authenticated as Plone's TEST_USER. But I don't see any hints of why this user should be allowed to add stuff.
Maybe you are missing a `self.grant()`? Maybe `Contributor`?

Reply:
Thanks a lot. I indeed did not set this permission. I did not inherit from the setup expecting it would be setup anyway.

Reply:
Great.
Our standard setup is usually like this:
- the current user is the Plone test user (`login()` is called in `plone.testing`)
- the user has no roles
- the browser is not logged in

Reply:
Yes, that's mostly why I did it wrong. I was used to login in the test which I copied from sl. But in this test the setup, which I did not copy, already logged in the user.